lib/src/component_declaration/function_component.dart

```diff
@@ -0,0 +1,188 @@
+///   Iterable<String> items;
+/// }
+/// ```
+///
+/// __OR__ Optionally pass in an existing [PropsFactory] in place of a props [config].
+///
+/// ```dart
+/// UiFactory<FooProps> Bar = uiFunction(
+///   (props) {
+///     return (Dom.button()..disabled = props.isDisabled)('Click me!');
+///   },
+///   FunctionComponentConfig(
+///     propsFactory: PropsFactory.fromUiFactory(Foo),
+///     displayName: 'Bar',
+///   ),
+/// );
+/// ```
+///
+/// __OR__ Set [config] to `null` when using `UiProps`.
+///
+/// ```dart
+/// UiFactory<UiProps> Foo = uiFunction(
+///   (props) {
+///     return Dom.div()('prop id: ${props.id}');
+///   },
+///   FunctionComponentConfig(
+///     displayName: 'Foo',
+///   ),
+/// );
+/// ```
+///
+/// Learn more: <https://reactjs.org/docs/components-and-props.html#function-and-class-components>.
+// TODO: right now only top level factory declarations will generate props configs.
+UiFactory<TProps> uiFunction<TProps extends UiProps>(
+  dynamic Function(TProps props) functionComponent,
+  FunctionComponentConfig<TProps> config) {
+  if (config == null) {
+    throw ArgumentError('config must not be null');
+  }
+
+  var propsFactory = config.propsFactory;
+
+  // Get the display name from the inner function if possible so it doesn't become `_uiFunctionWrapper`
+  final displayName = config.displayName ?? _getFunctionName(functionComponent);
+
+  dynamic _uiFunctionWrapper(JsBackedMap props) {
+    return functionComponent(propsFactory.jsMap(props));
+  }
+
+  final factory = react.registerFunctionComponent(
+    _uiFunctionWrapper,
+    displayName: displayName,
+  );
+
+  if (propsFactory == null) {
+    if (TProps != UiProps && TProps != _GenericUiProps) {
+      throw ArgumentError(
+          'config.propsFactory must be provided when using custom props classes');
+    }
+    propsFactory = PropsFactory.fromUiFactory(
+            ([backingMap]) => _GenericUiProps(factory, backingMap))
+        as PropsFactory<TProps>;
+  }
+
+  TProps _uiFactory([Map backingMap]) {
+    TProps builder;
+    if (backingMap == null) {
+      builder = propsFactory.jsMap(JsBackedMap());
+    } else if (backingMap is JsBackedMap) {
+      builder = propsFactory.jsMap(backingMap);
+    } else {
+      builder = propsFactory.map(backingMap);
+    }
+
+    return builder..componentFactory = factory;
+  }
+
+  return _uiFactory;
+}
+
+String _getFunctionName(Function function) {
+  return getProperty(function, 'name') ??
+      getProperty(function, '\$static_name');
+}
+
+class _GenericUiProps extends UiProps {
+  @override
+  final Map props;
+
+  _GenericUiProps(ReactComponentFactoryProxy componentFactory, [Map props])
+      : this.props = props ?? JsBackedMap() {
+    this.componentFactory = componentFactory;
+  }
+
+  @override
+  String get propKeyNamespace => '';
+
+  @override
+  bool get $isClassGenerated => true;
+}
+
+/// Helper class used to keep track of generated information for [uiFunction].
+@protected
```

Comment:
Now that we accept it as an argument to uiFunction, this class and its members shouldn't be `@protected` anymore